File: xkcd/src/main/java/xyz/jienan/xkcd/extra/fragment/SingleExtraWebViewFragment.java
package xyz.jienan.xkcd.extra.fragment;

import android.content.Intent;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.webkit.WebSettings;

import java.util.List;
import java.util.concurrent.TimeUnit;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import butterknife.BindDimen;
import butterknife.BindView;
import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import me.dkzwm.widget.srl.SmoothRefreshLayout;
import me.dkzwm.widget.srl.extra.footer.ClassicFooter;
import me.dkzwm.widget.srl.extra.header.ClassicHeader;
import timber.log.Timber;
import xyz.jienan.xkcd.R;
import xyz.jienan.xkcd.model.ExtraComics;
import xyz.jienan.xkcd.model.ExtraModel;
import xyz.jienan.xkcd.ui.RefreshFooterView;
import xyz.jienan.xkcd.ui.RefreshHeaderView;
import xyz.jienan.xkcd.whatif.fragment.SingleWhatIfFragment;

import static xyz.jienan.xkcd.Const.FIRE_EXTRA_SUFFIX;
import static xyz.jienan.xkcd.Const.FIRE_GO_EXTRA_MENU;
import static xyz.jienan.xkcd.Const.FIRE_SHARE_BAR;

/**
 * Created by jienanzhang on 03/03/2018.
 */

public class SingleExtraWebViewFragment extends SingleWhatIfFragment {

    @BindView(R.id.refresh_layout)
    SmoothRefreshLayout mRefreshLayout;

    private ExtraComics extraComics;

    private int currentPage = 0;

    @BindDimen(R.dimen.refresh_text_size)
    float refreshTextSize;

    public static SingleExtraWebViewFragment newInstance(ExtraComics extraComics) {
        SingleExtraWebViewFragment fragment = new SingleExtraWebViewFragment();
        Bundle args = new Bundle();
        args.putSerializable("entity", extraComics);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    protected int getLayoutResId() {
        return R.layout.fragment_extra_single;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Bundle args = getArguments();
        if (args != null)
            extraComics = (ExtraComics) args.getSerializable("entity");
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = super.onCreateView(inflater, container, savedInstanceState);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            webView.getSettings().setMixedContentMode(WebSettings.MIXED_CONTENT_ALWAYS_ALLOW);
        }
        webView.getSettings().setTextZoom((int) (sharedPref.getWhatIfZoom() * 1.5));
        if (savedInstanceState != null) {
            currentPage = savedInstanceState.getInt("current", 0);
        }
        loadLinkPage(currentPage);
        return view;
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        if (extraComics.links.size() > 1) {

            mRefreshLayout.setDisableLoadMore(false);
            mRefreshLayout.setDisablePerformRefresh(false);
            mRefreshLayout.setDisablePerformLoadMore(false);
            mRefreshLayout.setEnableKeepRefreshView(false);

            mRefreshLayout.setOnRefreshListener(new SmoothRefreshLayout.OnRefreshListener() {
                @Override
                public void onRefreshing() {
                    Timber.d("Refresh");
                    loadLinkPage(--currentPage);
                    mRefreshLayout.refreshComplete();
                    updateReleaseText();
                }

                @Override
                public void onLoadingMore() {
                    Timber.d("LoadMore");
                    loadLinkPage(++currentPage);
                    mRefreshLayout.refreshComplete();
                    Observable.timer(700, TimeUnit.MILLISECONDS)
                            .observeOn(AndroidSchedulers.mainThread())
                            .doOnNext(ignored -> webView.scrollTo(0, 0))
                            .subscribe();
                    updateReleaseText();
                }
            });
            mRefreshLayout.setEnableAutoRefresh(true);
            mRefreshLayout.setEnableAutoLoadMore(true);
            RefreshHeaderView header = new RefreshHeaderView(getContext());
            header.setTextSize(refreshTextSize);
            mRefreshLayout.setHeaderView(header);
            RefreshFooterView footer = new RefreshFooterView(getContext());
            footer.setTextSize(refreshTextSize);
            mRefreshLayout.setFooterView(footer);
            updateReleaseText();
        }
    }

    @Override
    public void onDestroyView() {
        if (compositeDisposable != null) {
            compositeDisposable.dispose();
        }
        if (webView != null) {
            webView.clearTasks();
        }
        super.onDestroyView();
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.action_share:
                Intent shareIntent = new Intent();
                shareIntent.setAction(Intent.ACTION_SEND);
                shareIntent.putExtra(Intent.EXTRA_TEXT, getString(R.string.share_text_article_extra, extraComics.links.get(0)));
                shareIntent.setType("text/plain");
                startActivity(Intent.createChooser(shareIntent, getResources().getText(R.string.share_to)));
                logUXEvent(FIRE_SHARE_BAR + FIRE_EXTRA_SUFFIX);
                return true;
            case R.id.action_go_explain: {
                Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(extraComics.explainUrl));
                startActivity(browserIntent);
                logUXEvent(FIRE_GO_EXTRA_MENU);
                return true;
            }
        }
        return super.onOptionsItemSelected(item);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("current", currentPage);
    }

    private void loadLinkPage(int pageIndex) {

        List<String> links = extraComics.links;

        compositeDisposable.add(
                ExtraModel.getInstance().parseContentFromUrl(links.get(Math.abs(pageIndex % links.size())))
                        .subscribe(html -> webView.loadDataWithBaseURL("file:///android_asset/.",
                                html, "text/html", "UTF-8", null),
                                Timber::e));
    }

    private void updateReleaseText() {
        if (extraComics.num == 1) {
            if (Math.abs(currentPage) % 2 != 0) {
                ((ClassicHeader) mRefreshLayout.getHeaderView()).setReleaseToRefreshRes(R.string.release_for_puzzle);
                ((ClassicFooter) mRefreshLayout.getFooterView()).setReleaseToLoadRes(R.string.release_for_puzzle);
            } else {
                ((ClassicHeader) mRefreshLayout.getHeaderView()).setReleaseToRefreshRes(R.string.release_for_solution);
                ((ClassicFooter) mRefreshLayout.getFooterView()).setReleaseToLoadRes(R.string.release_for_solution);
            }
        }
    }
}

File: xkcd/src/main/java/xyz/jienan/xkcd/extra/fragment/SingleExtraWebViewFragment.kt
package xyz.jienan.xkcd.extra.fragment

import android.content.Intent
import android.content.res.Configuration
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.view.MenuItem
import android.view.View
import android.webkit.WebSettings
import androidx.core.os.bundleOf
import kotlinx.android.synthetic.main.fragment_extra_single.*
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import timber.log.Timber
import xyz.jienan.xkcd.Const.*
import xyz.jienan.xkcd.R
import xyz.jienan.xkcd.model.ExtraComics
import xyz.jienan.xkcd.model.ExtraModel
import xyz.jienan.xkcd.model.util.appendCss
import xyz.jienan.xkcd.ui.getUiNightModeFlag
import xyz.jienan.xkcd.whatif.fragment.SingleWhatIfFragment
import kotlin.math.abs

/**
 * Created by jienanzhang on 03/03/2018.
 */

class SingleExtraWebViewFragment : SingleWhatIfFragment() {

    private val extraComics by lazy { arguments?.getSerializable("entity") as ExtraComics }

    private var currentPage = 0

    override val layoutResId = R.layout.fragment_extra_single

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        customWebViewSettings()

        if (savedInstanceState != null) {
            currentPage = savedInstanceState.getInt("current", 0)
        }
        loadLinkPage(currentPage)
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onDestroyView() {
        compositeDisposable.dispose()
        if (webView != null) {
            webView.clearTasks()
        }
        super.onDestroyView()
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        if ((parentFragment as ExtraMainFragment).currentIndex != 1) { // use 0 as we only have 1 extra WebView at first page
            return false
        }
        when (item.itemId) {
            R.id.action_share -> {
                val shareIntent = Intent()
                shareIntent.action = Intent.ACTION_SEND
                shareIntent.putExtra(Intent.EXTRA_TEXT, getString(R.string.share_text_article_extra, extraComics.links?.get(0)))
                shareIntent.type = "text/plain"
                startActivity(Intent.createChooser(shareIntent, resources.getText(R.string.share_to)))
                logUXEvent(FIRE_SHARE_BAR + FIRE_EXTRA_SUFFIX)
                return true
            }
            R.id.action_go_explain -> {
                val browserIntent = Intent(Intent.ACTION_VIEW, Uri.parse(extraComics.explainUrl))
                startActivity(browserIntent)
                logUXEvent(FIRE_GO_EXTRA_MENU)
                return true
            }
        }
        return super.onOptionsItemSelected(item)
    }

    override fun onSaveInstanceState(outState: Bundle) {
        super.onSaveInstanceState(outState)
        outState.putInt("current", currentPage)
    }

    private fun loadLinkPage(pageIndex: Int) {
        val links = extraComics.links
        if (links != null) {
            ExtraModel.parseContentFromUrl(links[abs(pageIndex % links.size)])
                    .map {
                        if (context?.getUiNightModeFlag() == Configuration.UI_MODE_NIGHT_YES) {
                            val doc = Jsoup.parse(it)
                            doc.head().appendCss("night_style.css")
                            doc.html()
                        } else {
                            it
                        }
                    }
                    .map {
                        // Check solution / Go back to puzzle,
                        val doc = Jsoup.parse(it)
                        val (text, uri) =  if (pageIndex == 0) {
                            getString(R.string.check_solution) to  URI_XKCD_EXTRA_SOLUTION
                        } else  {
                            getString(R.string.check_puzzle) to URI_XKCD_EXTRA_PUZZLE
                        }
                        doc.body().children().last().appendElement("br").appendChild(Element("a").appendText(text).attr("href", uri))
                        doc.html()
                    }
                    .subscribe({ html ->
                        webView.loadDataWithBaseURL("file:///android_asset/.",
                                html, "text/html", "UTF-8", null)

                        webView.addXkcdUriInterceptor { url ->
                            if (url == URI_XKCD_EXTRA_SOLUTION) {
                                currentPage = 1
                                loadLinkPage(currentPage)
                            } else if (url == URI_XKCD_EXTRA_PUZZLE) {
                                currentPage = 0
                                loadLinkPage(currentPage)
                            }
                        }
                    }, { Timber.e(it) })
                    .also { compositeDisposable.add(it) }
        }
    }

    private fun customWebViewSettings() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            webView.settings.mixedContentMode = WebSettings.MIXED_CONTENT_ALWAYS_ALLOW
        }
        webView.settings.textZoom = (sharedPref.whatIfZoom * 1.5).toInt()
        if (Build.VERSION.SDK_INT < Build.VERSION_CODES.KITKAT) {
            webView.settings.layoutAlgorithm = WebSettings.LayoutAlgorithm.SINGLE_COLUMN
        }
    }

    companion object {
        fun newInstance(extraComics: ExtraComics) =
                SingleExtraWebViewFragment().apply { arguments = bundleOf("entity" to extraComics) }
    }
}
